horovod/_keras/__init__.py

```diff
@@ -41,34 +53,129 @@ def get_gradients(self, loss, params):
+                    for idx, grad in enumerate(self.grads):
+                        if self._sparse_as_dense and \
+                                isinstance(grad, tf.IndexedSlices):
+                            grad = tf.convert_to_tensor(grad)
+                        grad_aggregation_variable_name = str(idx)
+                        grad_aggregator = tf.get_variable(
+                            grad_aggregation_variable_name)
+                        update_op = grad_aggregator.assign_add(grad)
+                        aggregation_ops_list.append(update_op)
+                return aggregation_ops_list
+
+            def allreduce_grads():
+                if self._aggregation_frequency > 1:
+                    # Read in latest variables values.
+                    aggregated_grads = []
+                    aggregation_read_ops_list = []
+                    with tf.variable_scope("aggregation_variables", reuse=True):
+                        for idx, grad in enumerate(self.gpu_shadow_vars):
+                            grad_aggregation_variable_name = str(idx)
+                            grad_aggregator = tf.get_variable(
+                                grad_aggregation_variable_name)
+                            aggregated_grads.append(
+                                grad_aggregator.read_value())
+                            aggregation_read_ops_list.append(
+                                aggregated_grads[idx])
+                    aggregation_read_ops = tf.group(
+                        *aggregation_read_ops_list)
+                else:
+                    aggregated_grads = self.grads
+                    aggregation_read_ops = tf.no_op()
+
+                with tf.control_dependencies([aggregation_read_ops]):
+                    averaged_gradients = []
+                    for idx, grad in enumerate(aggregated_grads):
+                        if grad is not None:
                             avg_grad = hvd.allreduce(grad,
                                                      device_dense=self._device_dense,
                                                      device_sparse=self._device_sparse,
                                                      compression=self._compression)
                             averaged_gradients.append(avg_grad)
                         else:
                             averaged_gradients.append(None)
-                    return averaged_gradients
+                    with tf.control_dependencies([g.op for g in averaged_gradients]):
+                        reset_op = self.counter.assign(
+                            tf.constant(0), use_locking=True)
+                    with tf.control_dependencies([reset_op]):
+                        return [tf.identity(g) for g in averaged_gradients]
+
+            self._get_gradients_used = True
+            self.grads = super(
+                self.__class__, self).get_gradients(loss, params)
+            init_aggregation_vars()
```

Comment:
Why doesn't this call re-initialize the variables every time ?

Reply:
Here's the best explanation I can come up with. `init_aggregation_vars` has no reference to it later in the function. So there's no dependency placed on it by anything else that sets up the graph. Everything else can be traced back from the return. So it will get called once when initializing the graph, but won't be called again ever again. This was verified by some prints.